### PR TITLE
[Client] avoid locking in async send

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
As @iycheng discovered in https://github.com/ray-project/ray/issues/22082#issuecomment-1031821631, when `ClientObjectRef` is being GC'ed, `DataClient.lock` is acquired which may cause deadlock. This change avoids acquiring lock in `DataClient._async_send()`.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/util/client/dataclient.py
Comment (ckw017) [inline python/ray/util/client/dataclient.py:56]: Why are we switching from `request.req_id` to `id(request)`

Comment (ckw017) [reply python/ray/util/client/dataclient.py:56]: Ah okay, I guess it's since req_id generation is now deferred depending on how far along request_queue is processed

File: python/ray/util/client/dataclient.py
Comment (iycheng) [inline python/ray/util/client/dataclient.py:314]: I think there is one thing different from before when the cluster is down.
Previously it'll throw an exception, but now, it'll be pushed to the queue. Is there anything we need to take care here?

File: python/ray/util/client/dataclient.py
Comment (mwtian) [reply python/ray/util/client/dataclient.py:56]: Exactly, this enables not touching req_id in async send.

File: python/ray/util/client/dataclient.py
Comment (mwtian) [reply python/ray/util/client/dataclient.py:314]: Which exception?

**Review by ckw017 — APPROVED** [review]
lgtm, let me know if you need help with the reconnect tests

File: python/ray/util/client/dataclient.py
Comment (iycheng) [reply python/ray/util/client/dataclient.py:314]: let's add some comments here. so that future maintainers won't make the same mistake.

File: python/ray/util/client/dataclient.py
Comment (iycheng) [inline python/ray/util/client/dataclient.py:90]: you should try catch here. maybe sleep for 1 and blocking.

Comment (mwtian) [reply python/ray/util/client/dataclient.py:90]: What kind of exceptions need to be catched here?

File: python/ray/util/client/dataclient.py
Comment (mwtian) [reply python/ray/util/client/dataclient.py:314]: Added a comment above `_check_shutdown()`. Let me know if anything else can be documented better.

File: python/ray/util/client/dataclient.py
Comment (iycheng) [reply python/ray/util/client/dataclient.py:91]: https://docs.python.org/3/library/queue.html#queue.Queue.get

I saw you check request None here. My point is that `get` will not return None.

File: python/ray/util/client/dataclient.py
Comment (iycheng) [reply python/ray/util/client/dataclient.py:337]: I mean comment like:
```
This function will be called from python GC and we should avoid using the lock to prevent a deadlock.
```

File: python/ray/util/client/dataclient.py
Comment (mwtian) [reply python/ray/util/client/dataclient.py:91]: `None` is a special sentinel to signal shutdown of the data client, sent during `close()`.

Comment (iycheng) [reply python/ray/util/client/dataclient.py:91]: Got it. Thanks for the explanation!

**Comment by iycheng** [comment]
LGTM.

**Comment by iycheng** [comment]
Maybe, as an alternative, we can merge this one https://github.com/ray-project/ray/pull/22187/files and cherry-pick it to HA branch to reduce the risk, and merge this PR to master after.

File: python/ray/util/client/dataclient.py
Comment (mwtian) [reply python/ray/util/client/dataclient.py:348]: Added. Btw I find it easier to raise an exception from _async_send() when the client has shut down.

**Comment by iycheng** [comment]
@mwtian  @ckw017  should we revert this one, it seems to make the windows test very flaky.

**Comment by mwtian** [comment]
I assume the only `windows://python/ray/serve:test_ray_client` is affected? Sent #22283.

**Pull request merged**
